File: experiments/BENEFICIAL_GENESIS_DESIGN_001/protocol/merkle.py
"""Synthetic Bitcoin-like transaction identity, headers, and Merkle proofs.

This is a bounded model for inclusion proofs. It is not a Bitcoin full node
and does not speak to mainnet consensus, reorg policy beyond fixtures, or
script execution.
"""

from __future__ import annotations

from typing import Any, Iterable, Sequence

from .constants import DOMAIN_HEADER, DOMAIN_MERKLE, DOMAIN_TXID
from .encoding import canonical_json_bytes, domain_hash, domain_hash_hex, require_hex, sha256


def txid_from_tx(tx: dict[str, Any]) -> str:
    """Compute synthetic txid from the canonical transaction object.

    Only the identity-relevant fields are hashed. Witness-like fields that are
    not part of the synthetic identity surface are excluded to model a stable
    txid under non-malleable encoding rules of this pack.
    """

    identity = {
        "inputs": tx.get("inputs"),
        "locktime": tx.get("locktime", 0),
        "outputs": tx.get("outputs"),
        "version": tx.get("version", 1),
    }
    return domain_hash_hex(DOMAIN_TXID, canonical_json_bytes(identity))


def merkle_parent(left: bytes, right: bytes) -> bytes:
    return domain_hash(DOMAIN_MERKLE, left, right)


def merkle_root(txids_hex: Sequence[str]) -> str:
    if not txids_hex:
        raise ValueError("empty merkle tree")
    level = [require_hex("txid", t, expected_bytes=32) for t in txids_hex]
    while len(level) > 1:
        if len(level) % 2 == 1:
            level.append(level[-1])
        nxt = []
        for i in range(0, len(level), 2):
            nxt.append(merkle_parent(level[i], level[i + 1]))
        level = nxt
    return level[0].hex()


def merkle_proof(txids_hex: Sequence[str], index: int) -> dict[str, Any]:
    if index < 0 or index >= len(txids_hex):
        raise IndexError("merkle index out of range")
    level = [require_hex("txid", t, expected_bytes=32) for t in txids_hex]
    branch: list[str] = []
    idx = index
    while len(level) > 1:
        if len(level) % 2 == 1:
            level.append(level[-1])
        if idx % 2 == 0:
            sibling = level[idx + 1]
        else:
            sibling = level[idx - 1]
        branch.append(sibling.hex())
        nxt = []
        for i in range(0, len(level), 2):
            nxt.append(merkle_parent(level[i], level[i + 1]))
        level = nxt
        idx //= 2
    return {
        "merkle_branch_hex": branch,
        "merkle_index": index,
        "merkle_root_hex": level[0].hex() if level else merkle_root(txids_hex),
    }


def verify_merkle_proof(
    txid_hex: str,
    merkle_root_hex: str,
    branch_hex: Sequence[str],
    index: int,
) -> bool:
    try:
        h = require_hex("txid", txid_hex, expected_bytes=32)
        root = require_hex("merkle_root", merkle_root_hex, expected_bytes=32)
        idx = index
        for sib_hex in branch_hex:
            sib = require_hex("branch", sib_hex, expected_bytes=32)
            if idx % 2 == 0:
                h = merkle_parent(h, sib)
            else:
                h = merkle_parent(sib, h)
            idx //= 2
        return h == root
    except (TypeError, ValueError):
        return False


def header_hash(header: dict[str, Any]) -> str:
    body = {
        "bits": header.get("bits", 0),
        "height": header["height"],
        "merkle_root_hex": header["merkle_root_hex"],
        "prev_hash_hex": header["prev_hash_hex"],
        "time": header.get("time", 0),
    }
    return domain_hash_hex(DOMAIN_HEADER, canonical_json_bytes(body))


def build_header_chain(
    *,
    start_height: int,
    prev_hash_hex: str,
    blocks: Iterable[dict[str, Any]],
) -> list[dict[str, Any]]:
    """Build a synthetic header chain. Each block needs merkle_root_hex and time."""

    chain: list[dict[str, Any]] = []
    height = start_height
    prev = prev_hash_hex
    for block in blocks:
        header = {
            "height": height,
            "prev_hash_hex": prev,
            "merkle_root_hex": block["merkle_root_hex"],
            "time": block.get("time", 1_700_000_000 + height),
            "bits": block.get("bits", 0x1D00FFFF),
        }
        hh = header_hash(header)
        header["header_hash_hex"] = hh
        header["txids_hex"] = list(block.get("txids_hex", []))
        chain.append(header)
        prev = hh
        height += 1
    return chain


def confirmations(inclusion_height: int, tip_height: int) -> int:
    if tip_height < inclusion_height:
        return 0
    return tip_height - inclusion_height + 1

File: experiments/BENEFICIAL_GENESIS_REPRODUCER_001/src/merkle.mjs
/**
 * Synthetic transaction identity, headers, Merkle proofs.
 */

import { DOMAIN_HEADER, DOMAIN_MERKLE, DOMAIN_TXID } from "./constants.mjs";
import {
  canonicalJsonBytes,
  domainHash,
  domainHashHex,
  requireHex,
} from "./encoding.mjs";

export function txidFromTx(tx) {
  const identity = {
    inputs: tx.inputs,
    locktime: tx.locktime ?? 0,
    outputs: tx.outputs,
    version: tx.version ?? 1,
  };
  return domainHashHex(DOMAIN_TXID, canonicalJsonBytes(identity));
}

export function merkleParent(left, right) {
  return domainHash(DOMAIN_MERKLE, left, right);
}

export function headerHash(header) {
  const body = {
    bits: header.bits ?? 0,
    height: header.height,
    merkle_root_hex: header.merkle_root_hex,
    prev_hash_hex: header.prev_hash_hex,
    time: header.time ?? 0,
  };
  return domainHashHex(DOMAIN_HEADER, canonicalJsonBytes(body));
}

export function confirmations(inclusionHeight, tipHeight) {
  if (tipHeight < inclusionHeight) return 0;
  return tipHeight - inclusionHeight + 1;
}

export function verifyMerkleProof(txidHex, merkleRootHex, branchHex, index) {
  try {
    if (typeof index !== "number" || !Number.isInteger(index) || index < 0) return false;
    if (!Array.isArray(branchHex) || branchHex.length > 64) return false;
    if (index >= 2 ** branchHex.length) return false;
    let h = requireHex("txid", txidHex, { expectedBytes: 32 });
    const root = requireHex("merkle_root", merkleRootHex, { expectedBytes: 32 });
    let idx = index;
    for (const sibHex of branchHex) {
      const sib = requireHex("branch", sibHex, { expectedBytes: 32 });
      if (idx % 2 === 0) h = merkleParent(h, sib);
      else h = merkleParent(sib, h);
      idx = Math.floor(idx / 2);
    }
    return h.equals(root);
  } catch {
    return false;
  }
}

export function merkleRoot(txidsHex) {
  if (!txidsHex || txidsHex.length === 0) throw new Error("empty merkle tree");
  let level = txidsHex.map((t) => requireHex("txid", t, { expectedBytes: 32 }));
  while (level.length > 1) {
    if (level.length % 2 === 1) level.push(level[level.length - 1]);
    const nxt = [];
    for (let i = 0; i < level.length; i += 2) {
      nxt.push(merkleParent(level[i], level[i + 1]));
    }
    level = nxt;
  }
  return level[0].toString("hex");
}
